frontend/catalyst/compiler.py

```diff
@@ -119,37 +121,116 @@ def get_lib_path(project, env):
     "--relocation-model=pic",
 ]
 
-linker = "c99"
 
-mlir_lib_path = get_lib_path("llvm", "MLIR_LIB_DIR")
-linker_common_flags = [
-    "-shared",
-    "-rdynamic",
-    f"-L{mlir_lib_path}",
-    "-Wl,-no-as-needed",
-    f"-Wl,-rpath,{mlir_lib_path}",
-]
-
-common_libs = [
-    "-lpthread",
-    "-lmlir_c_runner_utils",  # required for memref.copy
-]
-
-runtime_libs = [
-    "-lrt_backend",
-    "-lrt_capi",
-    *common_libs,
-]
+# pylint: disable=too-few-public-methods
+class CompilerDriver:
+    """Compiler Driver Interface
+    In order to avoid relying on a single compiler at run time and allow the user some flexibility,
+    this class defines a compiler resolution order where multiple known compilers are attempted.
+    The order is defined as follows:
+    1. A user specified compiler via the environment variable CATALYST_CC. It is expected that the
+        user provided compiler is flag compatilble with GCC/Clang.
+    2. clang: Priority is given to clang to maintain an LLVM toolchain through most of the process.
+    3. gcc: Usually configured to link with LD.
+    4. c99: Usually defaults to gcc, but no linker interface is specified.
+    5. c89: Usually defaults to gcc, but no linker interface is specified.
+    6. cc: Usually defaults to gcc, however POSIX states that it is deprecated.
+    """
 
-lrt_lib_path = get_lib_path("runtime", "RUNTIME_LIB_DIR")
-lrt_capi_path = os.path.join(lrt_lib_path, "capi")
-lrt_backend_path = os.path.join(lrt_lib_path, "backend")
-lightning_linker_flags = [
-    *linker_common_flags,
-    f"-L{lrt_capi_path}",
-    f"-L{lrt_backend_path}",
-    f"-Wl,-rpath,{lrt_capi_path}:{lrt_backend_path}",
-]
+    _default_fallback_compilers = ["clang", "gcc", "c99", "c89", "cc"]
+
+    @staticmethod
+    def _flags():
+        mlir_lib_path = get_lib_path("llvm", "MLIR_LIB_DIR")
+        lrt_lib_path = get_lib_path("runtime", "RUNTIME_LIB_DIR")
+        lrt_capi_path = os.path.join(lrt_lib_path, "capi")
+        lrt_backend_path = os.path.join(lrt_lib_path, "backend")
+
+        flags = [
+            "-Wno-unused-command-line-argument",
+            "-Wno-override-module",
+            "-shared",
+            "-rdynamic",
+            f"-L{mlir_lib_path}",
+            "-Wl,-no-as-needed",
+            f"-Wl,-rpath,{mlir_lib_path}",
+            f"-L{lrt_capi_path}",
+            f"-L{lrt_backend_path}",
+            f"-Wl,-rpath,{lrt_capi_path}:{lrt_backend_path}",
+            f"-L{lrt_capi_path}",
+            f"-L{lrt_backend_path}",
+            f"-Wl,-rpath,{lrt_capi_path}:{lrt_backend_path}",
+            "-lrt_backend",
+            "-lrt_capi",
+            "-lpthread",
+            "-lmlir_c_runner_utils",  # required for memref.copy
+        ]
+
+        return flags
+
+    @staticmethod
+    def _cfo(fallback_compilers):
```

Comment:
I'll just add compiler_fallback_order. If you have another suggestion, please comment.

Reply:
Well it seems like this function returns available compilers on the system, so something like get_available_compilers or similar would be nice.

Reply:
It checks for the existence of the user specified compiler. If it is available on the path, then place it at the top of the compiler fallback order. Otherwise it returns the default compiler fallback order.

Reply:
Might make sense to revisit this once https://github.com/PennyLaneAI/catalyst/pull/30#discussion_r1123683478 is addressed.